src/common/utils.lua

```diff
@@ -241,47 +233,78 @@ function utils.camelCase(str)
-		Target = msg.From,
-		Action = "Credit-Notice",
-		Recipient = msg.Recipient,
-		Quantity = tostring(1),
-	}
-	for tagName, tagValue in pairs(msg) do
-		-- Tags beginning with "X-" are forwarded
-		if string.sub(tagName, 1, 2) == "X-" then
-			notice[tagName] = tagValue
+function utils.indexOf(t, value)
+	for i, v in ipairs(t) do
+		if v == value then
+			return i
+		end
+	end
+	return -1
+end
+
+function utils.updateUserAssociations(antId, state, user)
+	-- remove previous associations
+	if USERS[user] and user ~= state.Owner and not utils.includes(user, state.Controllers) then
+		local antIndex = utils.indexOf(USERS[user], antId)
+		if antIndex then
+			table.remove(USERS[user], antIndex)
 		end
 	end
 
-	return notice
+	-- add new associations
+	if user == state.Owner or utils.includes(user, state.Controllers) then
+		if not USERS[user] then
+			USERS[user] = {}
+		end
+		if not utils.includes(antId, USERS[user]) then
+			table.insert(USERS[user], antId)
+		end
+	end
 end
 
-function utils.notices.debit(msg)
-	local notice = {
-		Target = msg.From,
-		Action = "Debit-Notice",
-		Recipient = msg.Recipient,
-		Quantity = tostring(1),
-	}
-	-- Add forwarded tags to the credit and debit notice messages
-	for tagName, tagValue in pairs(msg) do
-		-- Tags beginning with "X-" are forwarded
-		if string.sub(tagName, 1, 2) == "X-" then
-			notice[tagName] = tagValue
+function utils.updateAssociations(antId, state)
+	-- Remove previous associations for old owner and controllers
+	local previousOwner = ANTS[antId].Owner
+	local previousControllers = ANTS[antId].Controllers
+
+	if previousOwner and previousOwner ~= state.Owner and type(ANTS[antId]) == "table" then
+		utils.updateUserAssociations(antId, ANTS[antId], previousOwner)
+	end
+
+	for _, user in ipairs(previousControllers) do
+		if not utils.includes(user, state.Controllers) then
+			utils.updateUserAssociations(antId, ANTS[antId], user)
 		end
 	end
 
-	return notice
+	-- Add new associations for new owner and controllers
+	local users = { state.Owner }
+	for _, user in ipairs(state.Controllers) do
+		table.insert(users, user)
+	end
+
+	for _, user in ipairs(users) do
+		utils.updateUserAssociations(antId, state, user)
+	end
+
+	-- Update the ANTS table
+	ANTS[antId].Owner = state.Owner
+	ANTS[antId].Controllers = state.Controllers
 end
 
--- @param notices table
-function utils.notices.sendNotices(notices)
-	for _, notice in ipairs(notices) do
-		ao.send(notice)
+-- it is possible to register an ANT that does not respond to the state request, so we need to clean up the ANTS table
+function utils.cleanAnts(currentTime)
+	for antId, ant in pairs(ANTS) do
+		if
+			ant.RegisteredAt
+			and not ant.Owner
+			-- We check for both Owner and Controllers to be empty, as it is possible for an ANT to have an Owner (a renounced ant)
+			and not ant.Controllers
+			-- If we don't recieve the state within 30 minutes consider the ANT as invalid
+			and currentTime - ant.RegisteredAt > 1000 * 60 * 30
```

Comment:
nit: move this to a constant, and consider a separate handler that can be invoked for cleanups.